### PR TITLE
Add antithesis PoC workload

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

This testing workload should be updated constantly with the tip of `master`. By including the workload into avalanchego we'll unblock pushing docker images on every change.

## How this works

Migrates antithesis workload from legacy repo into avalanchego.

## How this was tested

Running locally against a single node network works as expected:
```
...
2024/03/01 12:37:42 confirmed X-chain transaction 2PzD4RbsRuv6TCcW5vAQSxHBtFqyg4dXTE2o2bWt9GbHWi5TDP on all nodes
2024/03/01 12:37:42 confirmed all X-chain UTXOs consumed by 2PzD4RbsRuv6TCcW5vAQSxHBtFqyg4dXTE2o2bWt9GbHWi5TDP are not present on http://localhost:9650
2024/03/01 12:37:42 confirmed all X-chain UTXOs consumed by 2PzD4RbsRuv6TCcW5vAQSxHBtFqyg4dXTE2o2bWt9GbHWi5TDP are not present on all nodes
2024/03/01 12:37:42 confirmed P-chain transaction 2quMhqc8ZLDLpFSxLNMhtzbJGyeXh7MtcCevSzMbnXqSLEuiC6 on http://localhost:9650
2024/03/01 12:37:42 confirmed P-chain transaction 2quMhqc8ZLDLpFSxLNMhtzbJGyeXh7MtcCevSzMbnXqSLEuiC6 on all nodes
2024/03/01 12:37:42 confirmed all P-chain UTXOs consumed by 2quMhqc8ZLDLpFSxLNMhtzbJGyeXh7MtcCevSzMbnXqSLEuiC6 are not present on http://localhost:9650
2024/03/01 12:37:42 confirmed all P-chain UTXOs consumed by 2quMhqc8ZLDLpFSxLNMhtzbJGyeXh7MtcCevSzMbnXqSLEuiC6 are not present on all nodes
2024/03/01 12:37:42 created new X-chain asset 4mrzuyrKWPzFW48hGUvD1SkWKBHkwF8DkEqkr7ftuCD6rxCfY in 104.50525ms
2024/03/01 12:37:42 confirmed X-chain transaction 4mrzuyrKWPzFW48hGUvD1SkWKBHkwF8DkEqkr7ftuCD6rxCfY on http://localhost:9650
2024/03/01 12:37:42 confirmed X-chain transaction 4mrzuyrKWPzFW48hGUvD1SkWKBHkwF8DkEqkr7ftuCD6rxCfY on all nodes
2024/03/01 12:37:42 confirmed all X-chain UTXOs consumed by 4mrzuyrKWPzFW48hGUvD1SkWKBHkwF8DkEqkr7ftuCD6rxCfY are not present on http://localhost:9650
2024/03/01 12:37:42 confirmed all X-chain UTXOs consumed by 4mrzuyrKWPzFW48hGUvD1SkWKBHkwF8DkEqkr7ftuCD6rxCfY are not present on all nodes
2024/03/01 12:37:42 wallet 0 executing flow 2
2024/03/01 12:37:42 created new X-chain asset xxE17VE8KZkdHaETA9F8qxsMG3zjsBGmWZvdGorDaqLwveW5B in 106.018208ms
2024/03/01 12:37:42 issued X-chain operation tx onSqYXLFSFGx8xXayBaqDaCt8zVAKDJRjR2xnbSfrwMCkLuR9 in 106.146ms
2024/03/01 12:37:42 confirmed X-chain transaction xxE17VE8KZkdHaETA9F8qxsMG3zjsBGmWZvdGorDaqLwveW5B on http://localhost:9650
2024/03/01 12:37:42 confirmed X-chain transaction xxE17VE8KZkdHaETA9F8qxsMG3zjsBGmWZvdGorDaqLwveW5B on all nodes
2024/03/01 12:37:42 confirmed all X-chain UTXOs consumed by xxE17VE8KZkdHaETA9F8qxsMG3zjsBGmWZvdGorDaqLwveW5B are not present on http://localhost:9650
2024/03/01 12:37:42 confirmed all X-chain UTXOs consumed by xxE17VE8KZkdHaETA9F8qxsMG3zjsBGmWZvdGorDaqLwveW5B are not present on all nodes
2024/03/01 12:37:42 confirmed X-chain transaction onSqYXLFSFGx8xXayBaqDaCt8zVAKDJRjR2xnbSfrwMCkLuR9 on http://localhost:9650
2024/03/01 12:37:42 confirmed X-chain transaction onSqYXLFSFGx8xXayBaqDaCt8zVAKDJRjR2xnbSfrwMCkLuR9 on all nodes
...
```

### CODE REVIEW

File: tests/antithesis/main.go [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+		log.Printf("skipping X-chain tx issuance due to insufficient balance: %d < %d", avaxBalance, neededBalance)
+		return
+	}
+
+	var (
+		owner                = w.makeOwner()
+		createAssetStartTime = time.Now()
+	)
+	createAssetTx, err := xWallet.IssueCreateAssetTx(
+		"HI",
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply tests/antithesis/main.go:311]: Correct.

File: tests/antithesis/main.go [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+}
+
+func (w *workload) issueXChainBaseTx(ctx context.Context) {
+	var (
+		xWallet  = w.wallet.X()
+		xBuilder = xWallet.Builder()
+	)
+	balances, err := xBuilder.GetFTBalance()
+	if err != nil {
+		log.Printf("failed to fetch X-chain balances: %s", err)
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply tests/antithesis/main.go:243]: Right now anything starting with "failed" will be picked up by the monitoring.

**Comment by StephenButtolph** [comment]
> Is what appears in this PR sufficient for consumption, or do we also need e.g. a dockerfile.

We will also need to add a docker file along with a github action that pushes the docker image to the repository.

File: tests/antithesis/main.go [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+		if s.Status != status.Committed {
+			log.Printf("failed to confirm P-chain transaction %s on %s: status == %s", txID, uri, s.Status)
+			return
+		}
+		log.Printf("confirmed P-chain transaction %s on %s", txID, uri)
+	}
+	log.Printf("confirmed P-chain transaction %s on all nodes", txID)
+}
+
+func (w *workload) verifyXChainTxConsumedUTXOs(ctx context.Context, tx *xtxs.Tx) {
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply tests/antithesis/main.go:571]: The workload is supposed to touch as much functionality as possible and perform integrity checks... What those integrity checks are... is up to us.

File: tests/antithesis/main.go [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+	addr, _ := w.addrs.Peek()
+	return secp256k1fx.OutputOwners{
+		Threshold: 1,
+		Addrs: []ids.ShortID{
+			addr,
+		},
+	}
+}
+
+func (w *workload) confirmXChainTx(ctx context.Context, tx *xtxs.Tx) {
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply tests/antithesis/main.go:535]: The wallet is only confirming the transaction against the node it is connected to. This confirms the transaction across all the nodes.

File: tests/antithesis/main.go [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+	if avaxBalance < neededBalance {
+		log.Printf("skipping X-chain tx issuance due to insufficient balance: %d < %d", avaxBalance, neededBalance)
+		return
+	}
+
+	var (
+		owner                = w.makeOwner()
+		createAssetStartTime = time.Now()
+	)
+	createAssetTx, err := xWallet.IssueCreateAssetTx(
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply tests/antithesis/main.go:360]: They are touching slightly different code paths... But we could probably avoid it and just keep the operation tx test

**Review by dhrubabasu — APPROVED** [review]

**Pull request merged**